### ISSUE TITLE
Track Jev/Laya decision-model evaluation for OpenChrome reliability and safety reviewers

### ISSUE THREAD

**Issue by shaun0927** [shaun0927/openchrome#1666]
## Summary

Continue the decision-model evaluation that started in Claude session `8f4f7598-99b4-44e8-98d4-b94e0188d1fa`.

The current evidence supports adding a bounded decision-review surface to OpenChrome, but it does **not** yet support claiming that TypeSafe Jev, Laya, or any other System 1 decision model improves OpenChrome in production.

Current local verdict:

- G0, G1, G2: closed locally.
- G3: open.
- Completion audit verdict: `goal_not_complete`.
- Commercial recommendation verdict: `do_not_ship_jev_claim_yet`.

## Current Evidence Summary

Local evidence artifacts:

- `artifacts/decision/G3/CLAUDE_RECONSTRUCTION.md`
- `artifacts/decision/G3/ANALYSIS_REPORT.md`
- `artifacts/decision/G3/DECISION_MEMO.md`
- `artifacts/decision/G3/IMPLEMENTATION_MAP.md`
- `artifacts/decision/G3/COMMERCIAL_RECOMMENDATIONS.md`
- `artifacts/decision/G3/COMPLETION_AUDIT.md`
- `artifacts/decision/G3/LEDGER_AUDIT.md`
- `artifacts/decision/G3/RUNBOOK.md`
- `artifacts/decision/G3/CLOSURE_INPUT_TEMPLATE.md`

Verification already run locally:

- `npm test -- --runInBand tests/decisions/schema.test.ts tests/decisions/eval.test.ts tests/pilot/decider/decider.test.ts`
- `npm run build:src`
- `npm run preflight:g3`
- `npm run ledger-audit:g3`

Results:

- Tests: 3 suites / 36 tests passed.
- Build: passed.
- Ledger audit: passed, missing evidence 0, bad status 0.
- Preflight: command passed, but `ok_to_close: false`.

## Recommended OpenChrome Adoption Sequence

1. **Outcome review for failed or suspicious interactions**
   - Trigger: deterministic `classifyOutcome` returns `SILENT_CLICK` or `WRONG_ELEMENT`.
   - Proposed model role: bounded reviewer, not replacement for deterministic success fast path.
   - Commercial value: reliability and explainable recovery for silent failures.
   - Current blocker: needs approval record, host answers, and non-skipped model column.

2. **Irreversible-action policy review**
   - Trigger: mutation/irreversible action before execution.
   - Proposed model role: additional reviewer while deterministic policy remains hard boundary.
   - Commercial value: auditability, safety, and enterprise compliance positioning.
   - Current blocker: needs trust-model approval, approval record, host answers, and non-skipped model column.

3. **Element-search telemetry before model routing**
   - Do not route element selection to Jev/Laya yet.
   - Current A evidence: 28 A element cases, 0 have AX match-level telemetry, 0 have CSS score telemetry.
   - Next step: record `input.view.history[].resolution` with `resolver`, `ax_match_level`, and `css_score`.

4. **Gate detection exclusion for current G3**
   - Current labeled corpus has 0 gate cases across `bot_check`, `login`, `paywall`, `two_factor`, and `none`.
   - Current recommendation: approve exclusion from G3 unless labeled gate cases are added.

## Remaining G3 Blockers

`npm run preflight:g3` currently reports these blockers:

- `threshold_freeze`: `element_search`, `gate_detection`
- `host_file_column`: no host answers file covers every request
- `approval_records`: no approval response file covers every request
- `jev_column`: model column is skipped/missing for all four insertion points

## Laya as a Possible TypeSafe/Jev API Alternative

Laya should be tracked as a possible alternative provider, but it should not be treated as a drop-in free hosted API yet.

Observed from current public pages:

- `https://laya.convaiinnovations.com/` describes Laya as an open, Apache-2.0, non-autoregressive System 1 decision model family.
- The site claims local/open-weight operation with "zero API subscription costs" and a `pip install laya` quickstart.
- Hugging Face lists `convaiinnovations/laya` with license `apache-2.0` and describes three checkpoints:
  - root English checkpoint
  - `multilingual`
  - `typed-decisions`
- The public docs emphasize self-hosted/open-weight use and a Hugging Face demo. This is different from a guaranteed free production API.

Implication for OpenChrome:

- Laya can reduce or remove dependency on TypeSafe/Vercel hosted API **if** OpenChrome can run it locally or on self-managed infrastructure.
- It is not cost-free in an absolute sense: local/GPU/CPU inference still has compute, memory, packaging, latency, and deployment costs.
- Treat it as a new provider candidate, not as proof that the Jev column is complete.

## Proposed Follow-Up Work

Add a provider-neutral System 1 decision-model interface and compare:

- TypeSafe Jev via Vercel AI Gateway, when key/trust approval exists.
- Laya local provider, using the `typed-decisions` checkpoint first.
- Existing deterministic regex/current heuristics.
- Host-file answers as a degradation/control column.

Suggested local tasks:

1. Add `laya` provider support behind an explicit opt-in environment flag.
2. Map G3 `DecisionQuestion` to Laya `choice` questions.
3. Keep `label` and `truth_hint` out of provider input.
4. Run the same G3 B split for `outcome_classification` and `irreversible_policy`.
5. Only claim improvement if insertion tables contain non-skipped Laya/Jev results and pass the existing preflight/approval gates.

## Do Not Claim Yet

- Do not claim TypeSafe Jev improves OpenChrome until a real non-skipped Jev column exists.
- Do not claim Laya improves OpenChrome until it runs on the same G3 evaluation cases.
- Do not claim free hosted API availability for Laya based only on the public Laya site. Current evidence supports open-weight/self-hosted use, not a guaranteed free production API.
- Do not implement element routing through any model until AX/CSS telemetry supports a frozen trigger.
- Do not claim gate handling from a zero-case gate corpus.

**Comment by shaun0927** [comment]
G3 Laya local update (2026-09-20)

Local hosting path was tested instead of requiring TypeSafe/Vercel credentials.

What changed locally:
- Added a new opt-in decision evaluator provider: laya-local.
- Provider runs the Python laya package locally through a JSONL worker and uses the typed-decisions checkpoint.
- Windows Hugging Face cache needed HF_HUB_DISABLE_SYMLINKS=1 because symlink creation failed without elevated privileges.
- State payloads are normalized to valid UTF-8 before Laya tokenization.

Evidence generated:
- artifacts/decision/G3/LAYA_LOCAL_SMOKE.md: package import, checkpoint load, and prediction succeeded on CPU.
- artifacts/decision/G3/eval-laya-local.json: laya-local evaluated 89 B-split cases, non-skipped.
- artifacts/decision/G3/insertion-tables.md now includes laya-local as a local-model column.
- artifacts/decision/G3/closure-preflight.md now treats the model column as satisfied for the two current candidate insertions only: outcome_classification and irreversible_policy.

Current metrics from the B-split run:
- laya-local overall: 89 evaluated, accuracy 0.3034.
- outcome_classification trigger subset: 18 cases, accuracy 0.2222, abstain 0.2778.
- irreversible_policy: 24 cases, accuracy 0.7917, abstain 0.0833.
- element_search remains blocked; the laya-local element subset records provider errors and must not be claimed as ready.
- gate_detection remains blocked because the current corpus has 0 gate cases.

Current preflight status:
- ok_to_close remains false.
- Failed checks: threshold_freeze, host_file_column, approval_records.
- decision_model_column now passes for outcome_classification and irreversible_policy.

Verification run:
- npm test -- --runInBand tests/decisions/schema.test.ts tests/decisions/eval.test.ts tests/pilot/decider/decider.test.ts: passed, 36 tests.
- npm run build:src: passed.
- LAYA_LOCAL_ENABLED=1 LAYA_PYTHON=<temp venv python> npm run evidence:g3: passed and produced non-skipped laya-local evidence.

Do not claim commercial improvement yet. The updated completion audit still reports goal_not_complete until host answers, approval records, and the element/gate threshold blockers are handled.

**Comment by shaun0927** [comment]
Follow-up update (2026-09-20)

Continued after the Laya local run and tightened the remaining G3 blocker evidence.

Changes made locally:
- Updated host answer validation so a missing --answers/HOST_DECISION_ANSWERS file records ok=false with finding: answers file not supplied.
- Updated approval validation so a missing --approvals/G3_APPROVALS file records ok=false with finding: approval file not supplied.
- Regenerated validation reports, closure preflight, completion audit, and HANDOFF.md.
- Updated HANDOFF.md to remove the stale jev_column blocker and record the current passing decision_model_column check from the laya-local evidence.

Current preflight remains unchanged in substance:
- ok_to_close: false.
- Passing: verification_status, decision_model_column.
- Failing: threshold_freeze, host_file_column, approval_records.

Verification:
- npm run validate-host:g3: exit 0, ok=false, finding=answers file not supplied.
- npm run validate-approval:g3: exit 0, ok=false, finding=approval file not supplied.
- npm run preflight:g3: ok_to_close=false with the same three blockers.
- npm test -- --runInBand tests/decisions/schema.test.ts tests/decisions/eval.test.ts tests/pilot/decider/decider.test.ts: passed, 36 tests.
- npm run build:src: passed.

This does not close G3. It only makes the missing external-input reports match the preflight semantics so future handoff readers do not misread request-only validation as coverage.

**Comment by shaun0927** [comment]
Laya/OpenChrome bounded learning architecture update from Codex session.

Implemented locally:
- Added a bounded learning namespace under `src/pilot/learning/*` for two initial classifier-style tasks: `irreversible_policy` and `outcome_failure_triage`.
- Added privacy-filtered shadow event construction and stores. Learning events redact URLs, credentials/tokens, email-like strings, and phone-like strings before becoming dataset candidates.
- Added dataset/holdout scaffolding and adapter registry scaffolding. Promotion modes are constrained to `off`, `shadow`, and `assist`; there is no autonomous model-authority path.
- Connected `decideWithBudget` to an optional `shadowLearning` hook. Default runtime behavior is unchanged; when configured, it records model-vs-deterministic decisions and actual result evidence while returning the same final answer path.
- Added G3 artifact generator `npm run learning-architecture:g3`, producing `artifacts/decision/G3/LEARNING_ARCHITECTURE.md` and `learning-architecture.json`. `npm run runbook:g3` now includes this artifact in the full local verification list.

Verification:
- `npm run learning-architecture:g3` passed.
- `npm test -- --runInBand tests/pilot/learning/learning.test.ts tests/pilot/decider/decider.test.ts tests/decisions/schema.test.ts tests/decisions/eval.test.ts` passed: 42 tests.
- `npm run build:src` passed.
- `npm run preflight:g3` still exits 0 and correctly reports `ok_to_close: false` with blockers: `threshold_freeze`, `host_file_column`, `approval_records`.

Current interpretation:
- Laya is now represented as a bounded shadow/assist reviewer path, not as a browser-control authority.
- No online per-user fine-tuning, consent flow, production adapter loading, or automatic promotion was implemented.
- This creates the practical foundation for future local Laya fine-tuning only after labeled, privacy-filtered, holdout-validated data exists.

**Comment by shaun0927** [comment]
Codex implementation update: bounded Laya/OpenChrome learning loop product path is now implemented locally.

Implemented:
- Added opt-in local learning settings. Default is disabled; local-only capture activates only with explicit configuration such as `OPENCHROME_LEARNING=1`.
- Connected learning to actual pilot contract runtime points:
  - critical irreversible-action hook records `irreversible_policy` events when enabled;
  - non-success post-action contract outcomes record `outcome_failure_triage` events when enabled;
  - disabled mode creates no learning JSONL file.
- Added privacy-safe JSONL event path. Runtime event state is minimized to action/verdict/evidence summaries and avoids raw DOM, screenshots, credentials, and form values.
- Added `openchrome learning` CLI action surface through the compiled CLI:
  - `status`
  - `export`
  - `validate`
  - `eval`
  - `finetune` scaffold
  - `registry-list`
  - `registry-promote`
- Added dataset export/validation pipeline: task closed-choice validation, unlabeled exclusion, non-redacted exclusion, label-source filter, invalid-choice exclusion, stable train/holdout split, and report output.
- Added file-based adapter registry plus promotion gate. `assist` promotion requires explicit eval evidence; there is no `authority` mode.
- Added conservative assist combiner: deterministic policy remains the hard boundary, model-only `allow` cannot weaken a stricter deterministic decision.
- Updated G3 artifacts: `LEARNING_ARCHITECTURE.md`, `HANDOFF.md`, and `RUNBOOK.md`.

Verification:
- `npm test -- --runInBand tests/pilot/learning/learning-runtime.test.ts tests/pilot/learning/learning.test.ts tests/pilot/learning/learning-pipeline.test.ts tests/pilot/learning/learning-cli.test.ts tests/pilot/decider/decider.test.ts tests/decisions/schema.test.ts tests/decisions/eval.test.ts` passed: 51 tests.
- `npm run build` passed.
- `node dist/index.js learning status --json` passed and shows learning disabled by default.
- Dist JSONL smoke passed: generated one local `irreversible_policy` event under `artifacts/decision/G3/learning-runtime-smoke/events.jsonl`; validation correctly reports it as `skipped_unlabeled: 1` until labeled.
- `npm run preflight:g3` still exits 0 with `ok_to_close: false` and the expected blockers: `threshold_freeze`, `host_file_column`, `approval_records`.

Intentional non-features:
- No automatic online fine-tuning.
- No per-user background fine-tuning.
- No remote upload.
- No model action authority.
- No automatic adapter promotion.

Remaining G3 closure work is unchanged: threshold freeze inputs, host answer file coverage, and approval records.

**Comment by shaun0927** [comment]
## 로컬 변경의 PR 분리·보완·병합 완료

최종 main: `43d99a01ea9a7abba080e20d1f142edafe4fa258`. 아래 5개 PR 모두 순서대로 병합했습니다.

로컬에 남아 있던 변경을 5개 PR로 분리하고, 원본 작업 트리는 그대로 보존한 별도 작업 트리에서 보완했습니다.

1. #1667 평가 기반: main 병합. 선택지 검증·abstain 정규화·호출 예산·외부 요청 timeout 보완.
2. #1668 Laya 로컬 평가 worker: main 병합. opt-in, 최대 20개 선택지, subprocess 오류/timeout/종료 처리.
3. #1669 로컬 이벤트: main 병합. disabled 기본값, local_only opt-in, task allowlist, 저장 시 허용 metadata만 재구성.
4. #1670 데이터셋/registry CLI: main 병합. 민감·unlabeled·invalid 레코드 제외, 상태별 stable split, 독립 예측 채점, adapter/dataset 해시 결합과 재평가 승격.
5. #1671 G3 운영 문서: main 병합. 추적 문서와 생성기 정리, 누락·빈 증거를 완료로 처리하지 않는 preflight.

검토에서 발견한 abstain 선택지 우회, 공개 저장 API의 top-level 민감 데이터 저장, sensitivity 플래그 레코드 export를 수정하고 재검토했습니다.

검증: 최종 main의 깨끗한 checkout에서 전체 TypeScript 빌드와 관련 통합 18 suites / 169 tests 통과. 설치용 wrapper status는 learning disabled 확인. preflight 보고서 생성 성공, ok_to_close=false 확인. 각 PR의 독립 gate review와 GitHub CI 통과 후 병합했으며 runtime 변경 PR의 Ubuntu/Windows 설치 및 실제 headless Chrome 검증도 통과했습니다.

추가 검증: 평가 스크립트 전체 typecheck, 실제 CLI eval→registry promotion, compiled API JSONL 생성, npm pack dry-run. 전체 저장소 Jest full-test는 이번 범위에서 실행하지 않았고 관련·runtime·CLI 회귀 검증을 실행했습니다.

### 아직 완료를 주장하지 않는 것

- 실제 모델 fine-tuning 실행과 자동 학습은 구현하지 않습니다. finetune은 blocked scaffold입니다.
- Laya provider는 오프라인 평가용이며 실제 브라우저 runtime에서 registry adapter를 자동 로드하지 않습니다.
- 모델 action authority는 없습니다. 사용량 증가만으로 모델이 자동 개선되지 않습니다.
- G3는 별도의 독립 라벨, 실측 모델 결과, 임계값 승인과 host/approval 기록이 필요합니다. fresh checkout preflight는 false입니다.
- npm 버전업·게시·재설치는 이 변경 병합과 별도입니다.

canonical 문서: docs/dev/LEARNING_ARCHITECTURE.md, RUNBOOK.md, HANDOFF.md, learning-pipeline.md.
